src/vm.rs

```diff
@@ -623,12 +705,43 @@ impl<'a, S: Storage, C: PevmChain> Vm<'a, S, C> {
                     }
 
                     // TODO: We should move this changed check to our read set like for account info?
-                    for (slot, value) in account.changed_storage_slots() {
-                        write_set.push((
-                            self.hasher
-                                .hash_one(MemoryLocation::Storage(*address, *slot)),
-                            MemoryValue::Storage(value.present_value),
-                        ));
+                    for (&slot, value) in account.changed_storage_slots() {
+                        let memory_location = MemoryLocation::Storage(*address, slot);
+                        let memory_location_hash = self.hasher.hash_one(memory_location.clone());
+                        match evm.db().lazy_strategy {
+                            LazyStrategy::None => {
+                                write_set.push((
+                                    memory_location_hash,
+                                    MemoryValue::Storage(value.present_value),
+                                ));
+                            }
+                            LazyStrategy::RawTransfer => unreachable!(),
+                            LazyStrategy::ERC20Transfer { .. } => {
+                                match Ord::cmp(&value.present_value, &value.original_value) {
+                                    std::cmp::Ordering::Less => {
+                                        lazy_locations.push(memory_location);
+                                        write_set.push((
+                                            memory_location_hash,
+                                            MemoryValue::ERC20TransferSender(
+                                                value.original_value - value.present_value,
+                                            ),
+                                        ));
+                                    }
+                                    std::cmp::Ordering::Equal => {
+                                        unreachable!();
```

Comment:
Minor but let's put all unreachables at the bottom.

Reply:
👍 I have added a comment to clarify:

```
Ordering::Equal => {
    // Unreachable because we are iterating `account.CHANGED_storage_slots()`.
    unreachable!();
}
```